T452: Implement 802.11ac functionality

alainlamar opened the pull request:
This improvement is old-style Vyatta compatible Perl and Bash. Rewrite will follow at a later stage. This improvement is fully backwards compatible, so existing setups don't need to have their configurations altered.

5GHz WiFi is now functional with those notes:

* Users must install compatible firmware for their cards if VyOS (Debian Jessie) does not have it in stock
* Users must know the capabilities and limitations of their cards (`iw phy phy0 list` may help)
* Users must check their regulatory domain settings when setting up their cards for the first times (`iw reg get`; `iw phy phy0 info`, etc) and make sure the channel mappings and emitted signal strength matches their country's laws
* Users may have to patch their regdb (MUST patch if based in Germany and using DFS)
* Users may have to edit `/etc/default/crda` to set their regulatory domain (example: `REGDOMAIN=DE`)
* Users may have to add `/etc/modprobe.d/cfg80211.cfg` and add an option for their regulatory domains (example: `options cfg80211 ieee80211_regdom=DE`) However, I'm planning an operations command that helps users set these values...

Please consider a merge...

alainlamar committed:
T452: Enable support for 5 GHz WiFi AccesPoints w/ DFS

alainlamar committed:
T452: Enable support for 5 GHz WiFi AccesPoints w/ DFS

alainlamar committed:
Merge branch '802.11ac' of https://github.com/alainlamar/vyatta-wireless into 802.11ac

alainlamar committed:
T452: Bugfix - country node was ill-referenced

alainlamar committed:
T452: Add mgmt-frame-protection config node (ieee80211w=[012])

alainlamar committed:
T452: dump_file is deprecated in hostapd 2.3-1+deb8u5, causes warning.

alainlamar committed:
T452: Fixed allowed channel ranges to match known usable channels, Refs:
https://en.wikipedia.org/wiki/List_of_WLAN_channels#5_GHz_(802.11a/h/j/n/ac)

alainlamar committed:
T452: Code commenting and cleaning

alainlamar committed:
T452: Add config node max-stations to set max_num_sta in hostapd.conf

alainlamar committed:
T452: Add config nodes to configure card capabilities and auto-set some
dependencies

alainlamar committed:
T452: Bugfix - When configuring channel 0, it would not be written to
hostapd config.

alainlamar committed:
T452: Add config node option capabilities vht MAX-A-MPDU-LEN-EXP3

alainlamar committed:
T452: Add config node reduce-transmit-power to configure hostapd
local_pwr_constraint

alainlamar committed:
T452: Add config node expunge-failing-stations to configure hostapd
disassoc_low_ack

alainlamar committed:
T452: Code cleaning, removed excess " from help phrases

alainlamar committed:
T452: Add config nodes vht-center-freq/freq-[12] to set channel index for multi channel
center frequencies

alainlamar committed:
T452: Bugfix - clear out misunderstanding about STBC referring to
number of spacial streams

alainlamar committed:
T452: Bugfix - Enable autocomplete

alainlamar committed:
T452: Bugfix - Enable autocomplete

alainlamar committed:
T452: Bugfix - Enable autocomplete

alainlamar committed:
T452: Bugfix - Syntax check was not written for boolean values but for characters which prevented wireless config from being loaded

alainlamar committed:
T452: Refactor capabilities ht node for easier use

alainlamar committed:
T452: Remove the god-mode hostapd-option config node

alainlamar committed:
T452: Add config node isolate-stations to implement STA isolation on the AP

alainlamar committed:
T452: Refactor capabilities vht node for easier use

alainlamar committed:
T452: Add config node ht auto-powersave since not all cards support AP-side u-APSD

alainlamar committed:
T452: Bugfix - config node ht auto-powersave did not work
T452: Bugfix - config node require-ht presence was not tested
T452: Bugfix - config node vht link-adaptation dit not work
T452: Bugfix - config node require-vht presence was not tested
T452: Bugfix - config node vht stbc rx indicated wrong value ranges

alainlamar committed:
T452: Bugfix - rewrite help text, antenna count will be automatically corrected if SU-BEAMFORMER is set.

alainlamar committed:
T452: Bugfix - rewrite help text, vht and ht tag nodes do more now than former multi nodes.

alainlamar committed:
T452: Code cleaning.

c-po commented:
Hi alain,

> Users must install compatible firmware for their cards if VyOS (Debian Jessie) does not have it in stock

Do we have sufficient fw binaries shipped in linux-firmware from Debian jessie? Or should we possibly ship our own linux-firmware package?

> Users must know the capabilities and limitations of their cards (iw phy phy0 list may help)

This sounds like a good point for a completion helper like the one to autocomplete interface names.

> Users may have to patch their regdb (MUST patch if based in Germany and using DFS)

Will it be legal to ship this patch in VyOS?

> Users may have to edit /etc/default/crda to set their regulatory domain (example: REGDOMAIN=de)

This is implemented in your second PR, right?

—
Christian

alainlamar commented:
Thanks for your input!

I remember Daniil or Yuri writing several months ago that `VyOS already
had too many packages`. Since I was unsure of
  
1. how many users would ever use the 5GHz WiFi subsystem and
2. if they really could not be made to download and install recent Debian firmware-<you-name-it> packages by themselves, 

I did not push this matter to daylight until sending the pull request. If changes were to be made, they needed to take place in the VyOS build scripts where stuff gets bundled into the squashfs on build time - put firmware packages from Debian stretch/testing there.

The makers of the Linux `iw` tool stated very clearly that their tool must not be `screen-scraped` as they did not consider its output stable. However, the old Vyatta wireless Perl scripts did not care about that and parsed `iw phy phyX info` output to test channel availability in context to the current regulatory domain. Also, the `iw` tool often indicates support for features the card may actually have, but the Kernel drivers would not support them - which you only learn at the moment where you wonder why hostapd would not start and issued a `/usr/sbin/hostapd -dd /var/run/hostapd/wlanX.cfg` just to see a string indicating the [Kernel] driver would not support one of the selected features yet. However, hostapd uses very sane defaults and every flag you find to be actually working will only add to the capabilities of your WiFi AP.

I'm no lawyer, but after reading several documents from Bundesnetzagentur, I don't see any reason why a VyOS build with the DFS Kernel and scripts to configure WiFi would be illegal. One would offend laws, if:

1. cards or antennas are used which are not certified by an European regulatory body (or in other cases, local authorities). Usually, the CE sign indicates the device creator did follow all regulations. Also, the Compex WLE600VX cards I used did undergo certification: https://www.compex.com.sg/product/wle600vx/#tab-certification.
2. radio equipment was configured and used outside regulatory domain settings valid for your country (radio emissions too strong, wrong protocols or frequencies, etc).

And last but not least, there are so many WiFi AP manufacturers selling their stuff all across Europe without putting any additional 5GHz exclusive stamps or certs on them compared to pure 2.4GHz equipment. While developing, I used an Android phone with a WiFi analyzer app (Fritz!App WLAN and WifiAnalyzer) to check over the air if my APs were coming up in the right channels. If really in doubt, just remove the AP antennas, so the WiFi signal emission will be very weak and detectable only from within a few meters proximity. I also used the VyOS Linux command line to install `wavemon` and print me the actual signal  emission strength, reg domain and channel number and width directly from the device.

Your are right with your last question. I wanted to abstract static regdom settings from the user as this is very easy and can be done almost fail safe. However, since there is still the config node `interfaces wireless wlanX country` for each card, things may get broken there. I'd like, at a later point in time and to not yet break backward compatibility, to move the `country` in front of the `wlanX` tag note so this single setting has effect on all WiFi cards. Upon commit, the commit script could then just call the operations script to make changes as needed. Also please note that calling `iw reg set DE` does not always work. Depending on your cards, drivers, firmware and other hiccups, the Kernel may silently refuse to change the effective regulatory domain, so you always have to inspect `dmesg` output and also call `iw reg get` to see if your WiFi drivers really adopted the settings.

Overall, I found it quite hard to get first-hand information from the Kernel about all relevant WiFi settings. Things were obfuscated and complicated by driver developers. They obviously were putting several booby traps into the WiFi subsystem to make it harder for anyone to just activate and use the full range of WiFi capabilities. You really need to get a firm grasp of how it works (aka look up the driver code, and necessarily the domain regulations) if you were to implement a hostapd configuration. I abstracted this as much as I could from the user, but without limiting the users to using exactly one card model within exactly one country, I cannot make everything just fail-save...

alainlamar committed:
T452: Remove dependency on Perl smartmatch module

alainlamar committed:
T452: WiFi regulatory domain settings are done in a single place using
      "set system wifi-regulatory-domain <country-code>".
      This replaces the ambiguous
      "set interfaces wireless wlanX country <country-code>" nodes.
      This commit depends on all prior commits in this branch.
      This feature will not make runtime changes to a
      VyOS wireless system. Static changes made need a VyOS reboot!
      !!! THIS WILL BREAK BACKWARDS COMPATIBILITY !!!

alainlamar committed:
T452: Added help pointing to workaround for bug T521

alainlamar committed:
T452: Bugfix - clarified that SSH may crash if DFS channels are chosen.
